### ISSUE TITLE
Whispers of a Machine fails on startup with "unresolved import 'Label::get_TextAlignment'"

### ISSUE THREAD

**Issue by i30817** [adventuregamestudio/ags#751]
In the ags3 branch.

```
No placeholder functions for the plugin 'agsspritefont' found. The game might fail to load.
No placeholder functions for the plugin 'agsteam' found. The game might fail to load.
Placeholder functions for the plugin 'agswadjetutil' found.
Loading game failed with error:
Script link failed.
Runtime error: unresolved import 'Label::get_TextAlignment'.

The game files may be incomplete, corrupt or from unsupported version of AGS.
```

Which i would think would be here https://github.com/adventuregamestudio/ags/blob/50f84a3eb6487e5d167a93b4d2038b10bcffced1/Engine/ac/label.cpp#L153

It's possible this is a ags4 game though (it's recent), and i'm not able to compile it atm (it fails linking with plugins) so i dunno. It would probably fail on the rest of the missing plugins anyway.

**Comment by ivan-mogilko** [comment]
> It's possible this is a ags4 game though (it's recent), 

The version it was compiled with should be reported in the beginning of the log.

But it may be easier to ask the authors of the game if they have used custom engine.

**Comment by i30817** [comment]
Ok
```

Adventure Game Studio v3.5 Interpreter
Copyright (c) 1999-2011 Chris Jones and 2011-2019 others
ACI version 3.5.0.10

```

**Comment by ivan-mogilko** [comment]
Ok, don't know about "Whispers", 3.3.0 did not have that function yet, but apparently there's a bug in 3.5.0 because the property is called Label.TextAlignment in the script, yet registered as Label.Alignment in the engine. I will have to fix this of course.

**Comment by rofl0r** [comment]
> I will have to fix this of course.

indeed, because the other contributors to this repo are just interested in their "use moar C++ features" crusade

**Comment by morganwillcock** [comment]
@rofl0r is this issue related to C++ feature changes?

**Comment by i30817** [comment]
@ivan-mogilko  that fixed that import but this game has further problems because in AGSSpriteFont plugin.
```
Script link failed.
Runtime error: unresolved import 'SetVariableSpriteFont'.
```

I don't really understand how that is supposed to work. Some other plugins register without the macro this uses and use the format `engine->register("NameofClass:NameofFunc^numberofargs", (void*) NameofFunc);` but the macro apparently ignores both NameofClass and numberofargs.

When i tried to replace the macro by the style above it didn't work to fix it. Indeed, trying it on another cpp plugin (AGSBlend) introduced other runtime failures, so i don't even. I was thinking it was a conversion from c to c++ that was bad, but these facts say it isn't and i should let the debugging to pros.

**Comment by i30817** [comment]
Ok I got it. That plugin is not built in, i'm supposed to build the .so and place it on the game dir.

Now to yet another plugin error, may edit this on the future.

**Comment by ivan-mogilko** [comment]
Hmm, some of these could be useful additions to general engine.

`Mouse::set_ControlEnabled` -- we have only get_ControlEnabled (that's a readonly property based of user settings in config);
`Game::SimulateKeyPress` -- this was users request for some time now.

PS. Still waiting for author's reply, hopefully their programmer(s) will agree to shed some light on this.

BTW, is not Artistic License requires noting changes to the engine? Or is it only in case of open source distribution?

**Comment by i30817** [comment]
I lied, i built some more no-ops.

Namely (i think, but i'm blocked on AGSteam.so not being recognized), the engine needs these new functions:

SetLineHeightAdjust <-probably in AGSSpriteFont, probably with a int (font) and (float or int) arg and void return
Mouse::set_ControlEnabled <- probably in Engine/ac/mouse.cpp, probably with a bool arg and void return
Game::SimulateKeyPress^1 <- probably in Engine/ac/game.cpp, probably with a int argument (SDL key event or something) and a void return
MobileShowAchievements <- 'somewhere' i stuffed it on game.cpp but it probably should be part of ags2clientstub or something. Void return and void args (maybe).

Many maybes because i haven't run it yet to see the crashes. If i can even debug them. 

Anyway:
Anyone knows the reason why https://github.com/monkey0506/ags2clientstub/commits/master doesn't get recognized as a valid external (.so) plugin in the game dir while libagsspritefont.so is? It's blocking my hackery.

**Comment by ivan-mogilko** [comment]
> Anyone knows the reason why https://github.com/monkey0506/ags2clientstub/commits/master doesn't get recognized as a valid external (.so) plugin in the game dir while libagsspritefont.so is? It's blocking my hackery. I tried naming it agssteam.so, agsteam.so, libags2clientstub.so etc.

Ah, of course, on linux engine wants "libXXX.so". Sorry, I completely forgot about this.
So it has to be libagsteam.so, perhaps, or similar.

**Comment by i30817** [comment]
Guh, that was stupid. It was 'libagsteam.so'. Onwards to the next error 'fixed' badly for me.

**Comment by i30817** [comment]
So i got it to start... with the font completely fucked.


After i did my hackery i noticed it was crashing on Commons/font/fonts.cpp

After i put in fake functions for the plugin functions misisng; this was crashing:
```
bool is_bitmap_font(size_t fontNumber)
{
    if (fontNumber >= fonts.size() || !fonts[fontNumber].Renderer)
        return false;
    IAGSFontRenderer2 * a = fonts[fontNumber].Renderer2->IsBitmapFont();
}
```
I turned it into this:
```
bool is_bitmap_font(size_t fontNumber)
{
    if (fontNumber >= fonts.size() || !fonts[fontNumber].Renderer)
        return false;
    IAGSFontRenderer2 * a = fonts[fontNumber].Renderer2;
    if(!a)
         return false;
    return a->IsBitmapFont();
}
```

It makes the font look awful for some reason; probably SetLineHeightAdjust was supposed to make a shadow? Who knows.

**Comment by ivan-mogilko** [comment]
I think there's actually a bug in is_bitmap_font and maybe some other nearby function which does not comply with font renderer plugun api so your fix will make sense. I will look closer into this.

**Comment by ivan-mogilko** [comment]
I did this: bbbfe375431d025db22b795e868dad5c27437105

**Comment by i30817** [comment]
I'm just curious because i'm not working on code to make the game function better now that i have it booting (i have no clue how anyway), is a function pointer in C/C++ ie( `NameofFunction` standalone) the same as `&NameofFunction`? Because both kinds get passed to the function registering the plugin functions in different plugins and it's wigging me out.

**Comment by onitake** [comment]
Oh, I didn't notice somebody was already working on this!

I patched out AGS, AGSSpriteFont myself and added stubs for the new agswadjetutil symbols, but I'm getting a blank screen on startup. Looks like you were more successful.

About function pointers: Yes, a bare function name is a valid rvalue for a function pointer. You don't have reference it explicitly. `NameOfFunction` and `&NameOfFunction` are equivalent.

**Comment by onitake** [comment]
I can post the (trivial) modifications I made to support the additional script functions, if you're interested.

**Comment by i30817** [comment]
Blank screen is probably because you didn't force opengl on start and didn't update to the latest ags3 branch here (that does that). I'd think the project would be very interested in the mods (and i'd be very interested in not having a fork for no reason that supports a game badly).

Before you do though, check how it looks. In here, the font looks like this:
![image](https://user-images.githubusercontent.com/757900/57336957-a6ed3880-711f-11e9-80a8-33d5afa70df9.png)

**Comment by onitake** [comment]
Haha, somebody went ahead and fixed the new TextAlignment properties already: cec93fe47595db71875c3b9a590ba88ffaf6e029
That wasn't there on Sunday. :smile: 

Here's my changes so far: https://github.com/onitake/ags/tree/compat/whispers-of-a-machine

I haven't figured out what SetLineHeightAdjust does yet, only that it takes 3 arguments. At least that's what disassembling Whisper.exe with r2 tells me. I can look further into that, your screenshot indicates it's important.

**Comment by ivan-mogilko** [comment]
I got in touch with programmer who did "Kathy Rain" and "Whispers" and he agreed to share the source but wants to clean it up first :).

**Comment by i30817** [comment]
Kathy Rain has the same font problem so the source should be helpful.

**Comment by i30817** [comment]
Kathy Rain has the same font problem so the source above should be helpful, though i'm starting to think it's a hack of agsspritefont considering the differences between these methods: 
https://github.com/josthas/KrusAGS/blob/master/Plugins/AGSSpriteFont/AGSSpriteFont/VariableWidthSpriteFontRenderer.cpp#L31
https://github.com/adventuregamestudio/ags/blob/ags3/Plugins/AGSSpriteFont/AGSSpriteFont/VariableWidthSpriteFont.cpp#L36


I'm going to try to build the AGSSpriteFont dll on the kathy rain repo above and place it on both games dirs to see if it fixes stuff.

**Comment by i30817** [comment]
Anyway i managed to fix the font by following what the KathyRain repo did for `VariableWidthSpriteFontRenderer::RenderText` and `VariableWidthSpriteFontRenderer::Draw`, but since it's a hack not sure about pushing it upstream. I don't even understand what's happening there. I *suspect* it's drawing in black, then creating a new font glyph with a shrunken width and height and drawing again for a black outline.

**Comment by ivan-mogilko** [comment]
AGS engine does not care where the functions are coming from, so if functions are not good for general use one could still create a plugin with extra functions specifically for these games.

**Comment by i30817** [comment]
> We should definitely figure out what was changed in AGSSpriteFont and maybe merge this back into the AGS core. This seems to be too useful to ignore. Or remove AGSSpriteFont and instead reference what https://github.com/josthas/KrusAGS does?

The problem here is that this appears to be a hack. If you look at code differences like this:

https://github.com/josthas/KrusAGS/blob/master/Plugins/AGSSpriteFont/AGSSpriteFont/VariableWidthSpriteFont.cpp#L169  (and those 'if(false)' later on the function)

compared to:

https://github.com/adventuregamestudio/ags/blob/ags3/Plugins/AGSSpriteFont/AGSSpriteFont/VariableWidthSpriteFont.cpp#L142

> AGS engine does not care where the functions are coming from, so if functions are not good for general use one could still create a plugin with extra functions specifically for these games.

I think that may be needed. I was thinking that we could distinguish the behavior based on the colour argument to Draw (it's not used on the original function for some reason), but i doubt people were consistent with it.

**Comment by josthas** [comment]
Hi all, here's the WoaM engine source:

Cleaned it up as best I could, but there may be some junk in there still:
https://github.com/josthas/WoaMAGS

This WoaM engine source was essentially added on top of the KR engine source linked further up (but they had to be kept separate for savegame compatibility reasons). The original source was was branched off some Wadjet Eye fork of 3.3.x that happened to be the most stable engine build on Mac and iOS at the time (2014 I believe). So, yeah, we're dealing with quite old AGS code. 

I wouldn't encourage merging anything from either of these sources into the main branch. The few useful changes I put in has been implemented in later engine versions anyway (like crude resolution auto-detection, dynamic/unlimited object properties) etc. What is left are just a bunch of really horrible project specific hacks accumulated over time, most of which are completely undocumented and quickly whipped together mid-crunch to solve one problem or another.

So yeah, to conclude, I'd only suggest using these sources to get KR and/or WoaM up and running on Linux natively.

**Comment by i30817** [comment]
Thanks for the source. I'll just build the plugins for each game and place it on their dirs myself, but i'd like it if opensource-latest AGS could have compatibility for the games through a replacement hack without that for those people that have no clue.

I'd suggest also opening issues for those features you felt you needed a hack for, even if they're on plugins. Hammering out a design for papercuts is nice for everyone.

**Comment by i30817** [comment]
Eh, actually the built from source plugin is actually crashing for the open source ags engine if built directly but i can build a replacement just by copying some stuff.

**Comment by ivan-mogilko** [comment]
> but i'd like it if opensource-latest AGS could have compatibility for the games through a replacement hack 

I disagree with this. Like I said before, I would NOT like to put hacks into the ags engine to comply with the custom engine changes others made. My opinion is that this engine is not meant to be universal runner, this is what tools like ScummVM are for.

**Comment by onitake** [comment]
`SimulateKeypress` is something that could go into core, IMHO. Allegro already offers the functionality and it may be generally useful.

As for agswadjetutil: I don't know about all the functionality this plugin has, but the platform integrations may be good to have in core. Implementations can be stubbed when inapplicable, and picked up by a platform plugin where useful. So, like a generic API for achievements and the like. When running in vanilla AGS, achievements could simply be stored to/loaded from disk, like my stub implementation of AGSteam does, for example.

**Comment by ivan-mogilko** [comment]
Yes, that may make sense; it could also be reasonable to have generic achievement API inside AGS itself and only use plugins to connect to the actual service like Steam. Although I would consult with game developers regarding achivement fallback behavior to learn what they'd prefer.

Anyways, it looks like this issue is becoming overloaded, maybe open separate issues per problem then (like one for SimulateKeyPress and so on)?

**Comment by ivan-mogilko** [comment]
I just made 3 additions which let you run "Kathy Rain" and "Whispers of a Machine":

* Game.SimulateKeyPress (98d55b4)
* setter for Mouse.ControlEnabled (937862f)
* few stubs for "agsspritefont" plugin.

Please note that at least "Kathy Rain" still does not work 100% as with original exe even if you have built spritefont plugin from our repository: the text's color is mismatched for some reason.

Problem is, that the game itself does not have this plugin along. And the provided custom engine sources only have spritefont stubs in it (at least that's what I found so far). Not sure what's the deal with this, maybe WoaM sources contain more answers but I haven't researched them through yet.

It's still possible to play KR and WoaM without spritefont plugin in theory, they will use default stock font instead.

**Comment by i30817** [comment]
The font can be corrected, (and i did it) by modifying libagsspritefont to do this: https://github.com/onitake/ags/pull/1

Although that commit doesn't handle the 'line offsets' hack that kathy rain has on libagssprite font, i suppose it could be done (both for it and that color modification) by using a define and compile time flag. It's only used in two places i recall in the game (a postcard near the end and computer screen (in the hospital i think).

**Comment by i30817** [comment]
Here is a save where you can see the line adjustment being used https://github.com/onitake/ags/pull/1#issuecomment-491530409

Click on the postcard. And spoilers.

**Comment by i30817** [comment]
Lol, with the compile from ags release-3.5.0 and my altered plugin the colors for the outline and the font are inverted (white outline outside black 'font'). Weird.

**Comment by i30817** [comment]
Ok, i think the plugins can't be loaded for some reason on a read-only environment? I copied the game out of the squash filesystem and it worked to load the plugin(s) (i also had the galaxy stub from before it was built in).

**Comment by rofl0r** [comment]
fwiw when i played kathy rain i got the font issues fixed by simply unpacking the game files and scripts, reassemble a single file (e.g. globalscript.s), injecting it again into game28.dta and repacking it, then using that pack to play (while temporarily removing the original .exe due to #811). what i did to get it running is documented in the agsutils wiki.

**Comment by i30817** [comment]
Actually 

> Here is a save where you can see the line adjustment being used [onitake#1 (comment)](https://github.com/onitake/ags/pull/1#issuecomment-491530409)
> 
> Click on the postcard. And spoilers.

Actually this is fixed in your version of the plugin...

**Comment by i30817** [comment]
BTW @rofl0r , you can avoid the Kathy Rain bug in the parallax scroller bug by changing acsetup.cfg to add:
```
[override]
os=win
```

This disables the version of the module doing nothing in the 'not windows' branch. I think a new updated version of the module fixes this, FWIW if you want to show/plan how to fix this 'the right way'.

**Comment by ivan-mogilko** [comment]
Sorry, there's too much information on various issues here, and I am not sure which refers to me and this particular ticket.

Can we define on what is required for this ticket to be closed as fixed?

> Whispers of a machine still needs other things " 'MobileShowAchievements' and 'MobileGetAchievement'."

Ugh, I was considering adjusting script loading to not fail at first link error but gather them all from all modules and displaying in a list, although that would not work if they are in rooms (because engine does not want to load all existing room scripts at startup).

**Comment by onitake** [comment]
Please have a look at this branch: https://github.com/onitake/ags/pull/3

I'm still trying to wrap my head around all the changes that KrusAGS introduced...

**Comment by i30817** [comment]
My fear with those changes is: won't that make the plugin code incompatible with *other* games that haven't forked it?

Granted without the actual built plugin in the dir, AGS currently is just replacing all the functions by stubs, so the code would (probably) be broken anyway (if running), but if, for instance, that color change is merged, people building AGSSpritefont dll for game compatibility will end up with the fork.

Is the plan here a define? Patch the game at runtime to get rid of the incompatible stuff?

**Comment by onitake** [comment]
I don't think a #define is the right way, because it will cause further confusion on which fork of the plugin was used.

**Comment by onitake** [comment]
If you can, please try the modified plugin on several games and report how it goes.
I will do what I can to make it as backwards compatible as possible. If that requires removing the colour blending code, so be it.

**Comment by ivan-mogilko** [comment]
> For example, I suggest creating a separate ticket for spritefont plugin's differences.

Please?

**Comment by i30817** [comment]
Right. If you want to, feel free to close this and open a ticket for the functional differences. I'm satifisfied with the stubs (if those two for WoaM are added) because they allow to start the games and get them functional with the plugin prs on the onitake fork.

I don't think it's likely the fork will be merged to the upstream plugin, but that issue should be the place to sort it out.

**Comment by ivan-mogilko** [comment]
>  So basically, real plugin has the font problem + this line height problem, but shows the fridge puzzle (enter door, look at red fridge in the save); current stubs show a better font, but with the colors of of the outline reversed, and with the fridge puzzle broken

What is this "fridge puzzle", is this something logical that is broken? If it's a different thing, could you also open a new ticket for it?

**Comment by i30817** [comment]
It's a end game puzzle that modifies a font with the functionality of the original plugin (not the fork), thus it works if you built it, and doesn't with the stubs (if you do not have the dll on the dir).

**Comment by ivan-mogilko** [comment]
>  So it's not really 'broken', imo, unless you plan to make the stubs functional.

Hmm, no stubs are stubs, I'd rather wish there was a working open sourced plugin.

**Comment by i30817** [comment]
The funny thing is that i don't think the games actually make much use of the hacks the plugin fork made. My PR is very minimal and gets KR and WoaM font to display well, and there is only one thing broken AFAICT, that postcard, and i can hardcode a font to avoid it (unfortunately, i do no know how to tell from the plugin code if the game that is running is kathy rain, so i'm hesitant to do it).

Onitake wants to pull in all changes in sync with upstream, but i kind of suspect that is more trouble than it's worth, especially if they can't/won't be merged.

**Comment by onitake** [comment]
I refuse the idea of hardcoding per-game hacks.
Once you start going down that road, it will never end.

So, before this deteriorates further, I agree that we should open separate issues about all the little things were discussed in this thread. The AGSSpriteFont plugin is one of them.
Since I already fused the modifications off into a separate branch, I'm going to retarget against the main AGS3 repo.

**Comment by ivan-mogilko** [comment]
> Whispers of a machine still needs other things " 'MobileShowAchievements' and 'MobileGetAchievement'."

Please try d5bc68d3aa2a31ff5e840558ea7f299989b366f1

**Comment by i30817** [comment]
It worked.
